spring-cloud-stream-binder-kinesis/src/main/java/org/springframework/cloud/stream/binder/kinesis/KinesisMessageChannelBinder.java

```diff
@@ -146,21 +154,42 @@ protected MessageHandler createProducerMessageHandler(ProducerDestination destin
 			ExtendedProducerProperties<KinesisProducerProperties> producerProperties,
 			MessageChannel errorChannel) {
 
-		KinesisMessageHandler kinesisMessageHandler = new KinesisMessageHandler(
-				this.amazonKinesis);
-		kinesisMessageHandler.setSync(producerProperties.getExtension().isSync());
-		kinesisMessageHandler
-				.setSendTimeout(producerProperties.getExtension().getSendTimeout());
-		kinesisMessageHandler.setStream(destination.getName());
-		kinesisMessageHandler.setPartitionKeyExpression(
-				new FunctionExpression<Message<?>>((m) ->
-						m.getHeaders().containsKey(BinderHeaders.PARTITION_HEADER)
-								? m.getHeaders().get(BinderHeaders.PARTITION_HEADER)
-								: m.getPayload().hashCode()));
-		kinesisMessageHandler.setFailureChannel(errorChannel);
-		kinesisMessageHandler.setBeanFactory(getBeanFactory());
-
-		return kinesisMessageHandler;
+		FunctionExpression<Message<?>> partitionKeyExpression = new FunctionExpression<Message<?>>((m) ->
+			m.getHeaders().containsKey(BinderHeaders.PARTITION_HEADER)
+				? m.getHeaders().get(BinderHeaders.PARTITION_HEADER)
+				: m.getPayload().hashCode());
+		final AbstractAwsMessageHandler<?> messageHandler;
+		if (this.configurationProperties.isKplKclEnabled()) {
+			messageHandler = createKplMessageHandler(destination, partitionKeyExpression);
+		}
+		else {
+			messageHandler = createKinesisMessageHandler(destination, partitionKeyExpression);
+		}
+		messageHandler.setSync(producerProperties.getExtension().isSync());
+		messageHandler.setSendTimeout(producerProperties.getExtension().getSendTimeout());
+		messageHandler.setFailureChannel(errorChannel);
+		messageHandler.setBeanFactory(getBeanFactory());
+		return messageHandler;
+	}
+
+	private AbstractAwsMessageHandler<?> createKinesisMessageHandler(ProducerDestination destination,
+			FunctionExpression<Message<?>> partitionKeyExpression) {
+		final KinesisMessageHandler messageHandler;
+		messageHandler = new KinesisMessageHandler(this.amazonKinesis);
+		messageHandler.setStream(destination.getName());
+		messageHandler.setPartitionKeyExpression(partitionKeyExpression);
+		return messageHandler;
+	}
+
+	private AbstractAwsMessageHandler<?> createKplMessageHandler(ProducerDestination destination,
+			FunctionExpression<Message<?>> partitionKeyExpression) {
+		final KplMessageHandler messageHandler;
+		KinesisProducerConfiguration configuration = new KinesisProducerConfiguration();
+		configuration.setRegion(this.configurationProperties.getRegion());
```

Comment:
Would be better to populated a `setCredentialsProvider()` as which is provided by the auto-configuration from Spring Cloud AWS.
I think the same could be done for the `region` via `RegionProvider` from the same auto-configuration mentioned.

Also it looks like would be better to extract this `KinesisProducerConfiguration` into a bean which could be overridden by end-user. I see there are many configuration options could be provided...

Reply:
For the configuration options available, we put a comment about it on the spring-integration-aws merge request (https://github.com/spring-projects/spring-integration-aws/pull/103), which we should have written here instead, not sure if you saw it :

In the binder, there are a lot of parameters available in the `KinesisProducerConfiguration `(for aggregation / timeouts / metrics ...), but we do not use them. We only retrieve the region from the binder configuration.
The AWS KCL worker used in the adapter has also a lot of configuration parameters available in the `KinesisClientLibConfiguration`, we implement the mapping from the binder for the consumerGroup, the region, the shardIteratorType, the IdleBetweenPolls and the ConsumerBackoff.

Extracting the configuration for the producer will enable end users to override those properties, but there is no such bean for the kcl part. We can add support for other properties using the binder configuration.

Reply:
That's correct. Let's consider to expose more options for KCL/KPL in the separate issue here on the binder level!